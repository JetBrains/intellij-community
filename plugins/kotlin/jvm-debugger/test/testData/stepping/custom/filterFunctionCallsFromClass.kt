package filterFunctionCallsFromClass

class A {
    fun foo(): A {
        return this
    }

    private fun foo(i: Int, j: Int): A {
        return this
    }

    fun bar(): A {
        return this
    }

    private fun baz(): A {
        return this
    }

    inline fun inlineFoo(): A {
        return this
    }

    private inline fun inlineFoo(i: Int, j: Int): A {
        return this
    }

    inline fun inlineBar(): A {
        return this
    }

    private inline fun inlineBaz(): A {
        return this
    }

    fun testNoInlineCalls() {
        // STEP_OVER: 1
        //Breakpoint!
        stopHere()

        // SMART_STEP_INTO_BY_INDEX: 1
        // STEP_OUT: 1
        // SMART_STEP_TARGETS_EXPECTED_NUMBER: 3
        foo().bar().foo(1, 2).baz()

        // STEP_OVER: 1
        //Breakpoint!
        stopHere()

        // SMART_STEP_INTO_BY_INDEX: 2
        // STEP_OUT: 1
        // SMART_STEP_TARGETS_EXPECTED_NUMBER: 2
        foo().bar().foo(1, 2).baz()

        // STEP_OVER: 1
        //Breakpoint!
        stopHere()

        // SMART_STEP_INTO_BY_INDEX: 3
        // STEP_OUT: 1
        // SMART_STEP_TARGETS_EXPECTED_NUMBER: 1
        foo().bar().foo(1, 2).baz()
    }

    fun testFirstCallIsInline() {
        // STEP_OVER: 1
        //Breakpoint!
        stopHere()

        // SMART_STEP_INTO_BY_INDEX: 1
        // STEP_OUT: 1
        // SMART_STEP_TARGETS_EXPECTED_NUMBER: 3
        inlineFoo().bar().foo(1, 2).baz()

        // STEP_OVER: 1
        //Breakpoint!
        stopHere()

        // SMART_STEP_INTO_BY_INDEX: 2
        // STEP_OUT: 1
        // SMART_STEP_TARGETS_EXPECTED_NUMBER: 2
        inlineFoo().bar().foo(1, 2).baz()

        // STEP_OVER: 1
        //Breakpoint!
        stopHere()

        // SMART_STEP_INTO_BY_INDEX: 3
        // STEP_OUT: 1
        // SMART_STEP_TARGETS_EXPECTED_NUMBER: 1
        inlineFoo().bar().foo(1, 2).baz()
    }

    fun testFirstAndThirdCallsAreInline() {
        // STEP_OVER: 1
        //Breakpoint!
        stopHere()

        // SMART_STEP_INTO_BY_INDEX: 1
        // STEP_OUT: 1
        // SMART_STEP_TARGETS_EXPECTED_NUMBER: 3
        inlineFoo().bar().inlineFoo(1, 2).baz()

        // STEP_OVER: 1
        //Breakpoint!
        stopHere()

        // SMART_STEP_INTO_BY_INDEX: 2
        // STEP_OUT: 1
        // SMART_STEP_TARGETS_EXPECTED_NUMBER: 2
        inlineFoo().bar().inlineFoo(1, 2).baz()

        // STEP_OVER: 1
        //Breakpoint!
        stopHere()

        // SMART_STEP_INTO_BY_INDEX: 3
        // STEP_OUT: 1
        // SMART_STEP_TARGETS_EXPECTED_NUMBER: 1
        inlineFoo().bar().inlineFoo(1, 2).baz()
    }

    fun testAllCallsAreInline() {
        // STEP_OVER: 1
        //Breakpoint!
        stopHere()

        // SMART_STEP_INTO_BY_INDEX: 1
        // STEP_OUT: 1
        // SMART_STEP_TARGETS_EXPECTED_NUMBER: 3
        inlineFoo().inlineBar().inlineFoo(1, 2).inlineBaz()

        // STEP_OVER: 1
        //Breakpoint!
        stopHere()

        // SMART_STEP_INTO_BY_INDEX: 2
        // STEP_OUT: 1
        // SMART_STEP_TARGETS_EXPECTED_NUMBER: 2
        inlineFoo().inlineBar().inlineFoo(1, 2).inlineBaz()

        // STEP_OVER: 1
        //Breakpoint!
        stopHere()

        // SMART_STEP_INTO_BY_INDEX: 3
        // STEP_OUT: 1
        // SMART_STEP_TARGETS_EXPECTED_NUMBER: 1
        inlineFoo().inlineBar().inlineFoo(1, 2).inlineBaz()
    }
}

fun main() {
    val a = A()
    a.testNoInlineCalls()
    a.testFirstCallIsInline()
    a.testFirstAndThirdCallsAreInline()
    a.testAllCallsAreInline()
}

fun stopHere() {

}

// IGNORE_K2