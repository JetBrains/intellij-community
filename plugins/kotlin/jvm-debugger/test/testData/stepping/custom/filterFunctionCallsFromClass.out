LineBreakpoint created at filterFunctionCallsFromClass.kt:39
LineBreakpoint created at filterFunctionCallsFromClass.kt:48
LineBreakpoint created at filterFunctionCallsFromClass.kt:57
LineBreakpoint created at filterFunctionCallsFromClass.kt:68
LineBreakpoint created at filterFunctionCallsFromClass.kt:77
LineBreakpoint created at filterFunctionCallsFromClass.kt:86
LineBreakpoint created at filterFunctionCallsFromClass.kt:97
LineBreakpoint created at filterFunctionCallsFromClass.kt:106
LineBreakpoint created at filterFunctionCallsFromClass.kt:115
LineBreakpoint created at filterFunctionCallsFromClass.kt:126
LineBreakpoint created at filterFunctionCallsFromClass.kt:135
LineBreakpoint created at filterFunctionCallsFromClass.kt:144
Run Java
Connected to the target VM
filterFunctionCallsFromClass.kt:39
filterFunctionCallsFromClass.kt:44
filterFunctionCallsFromClass.kt:5
filterFunctionCallsFromClass.kt:44
filterFunctionCallsFromClass.kt:48
filterFunctionCallsFromClass.kt:53
filterFunctionCallsFromClass.kt:13
filterFunctionCallsFromClass.kt:53
filterFunctionCallsFromClass.kt:57
filterFunctionCallsFromClass.kt:62
filterFunctionCallsFromClass.kt:9
filterFunctionCallsFromClass.kt:62
filterFunctionCallsFromClass.kt:68
filterFunctionCallsFromClass.kt:73
filterFunctionCallsFromClass.kt:21
filterFunctionCallsFromClass.kt:73
filterFunctionCallsFromClass.kt:77
filterFunctionCallsFromClass.kt:82
filterFunctionCallsFromClass.kt:13
filterFunctionCallsFromClass.kt:82
filterFunctionCallsFromClass.kt:86
filterFunctionCallsFromClass.kt:91
filterFunctionCallsFromClass.kt:9
filterFunctionCallsFromClass.kt:91
filterFunctionCallsFromClass.kt:97
filterFunctionCallsFromClass.kt:102
filterFunctionCallsFromClass.kt:21
filterFunctionCallsFromClass.kt:102
filterFunctionCallsFromClass.kt:106
filterFunctionCallsFromClass.kt:111
filterFunctionCallsFromClass.kt:13
filterFunctionCallsFromClass.kt:111
filterFunctionCallsFromClass.kt:115
filterFunctionCallsFromClass.kt:120
filterFunctionCallsFromClass.kt:25
filterFunctionCallsFromClass.kt:120
filterFunctionCallsFromClass.kt:126
filterFunctionCallsFromClass.kt:131
filterFunctionCallsFromClass.kt:21
filterFunctionCallsFromClass.kt:131
filterFunctionCallsFromClass.kt:135
filterFunctionCallsFromClass.kt:140
filterFunctionCallsFromClass.kt:29
filterFunctionCallsFromClass.kt:140
filterFunctionCallsFromClass.kt:144
filterFunctionCallsFromClass.kt:149
filterFunctionCallsFromClass.kt:25
filterFunctionCallsFromClass.kt:149
Disconnected from the target VM

Process finished with exit code 0
